### PR TITLE
Cosocket num2str

### PR DESCRIPTION
**Pull request by zhuizhuhaomeng**
optimize: avoided use of lua_tolstring in ngx_http_lua_calc_strlen_in_table and ngx_http_lua_copy_str_in_table.
optimize: avoided use of lua_tolstring in ngx_http_lua_socket_udp_send, log_wrapper and ngx_http_lua_ngx_echo.

I hereby granted the copyright of the changes in this pull request
to the authors of this lua-nginx-module project.

### CODE REVIEW

File: src/ngx_http_lua_output.c [hunk 1]
```diff
@@ -245,7 +274,9 @@ ngx_http_lua_ngx_echo(lua_State *L, unsigned newline)
     return 1;
 }
 
-
```
Comment (doujiang24) [inline src/ngx_http_lua_output.c:248]: style: need keep two blank lines between two functions.

File: src/ngx_http_lua_output.c [hunk 2]
```diff
@@ -245,7 +274,9 @@ ngx_http_lua_ngx_echo(lua_State *L, unsigned newline)
...
+/**
+ * Get the maximum possible length, not the actual length
```
Comment (doujiang24) [inline src/ngx_http_lua_output.c:278]: it's better to move this comment to line 343

File: t/009-log.t [hunk 3]
```diff
@@ -547,4 +547,4 @@ ok
 --- no_error_log
 [error]
 --- error_log eval
-"2: hello\0world, client: "
+"2: hello\0world, client: "
```
Comment (doujiang24) [inline t/009-log.t:550]: unexpected change from the editor?

File: src/ngx_http_lua_output.c [hunk 4]
```diff
@@ -93,6 +95,15 @@ ngx_http_lua_ngx_echo(lua_State *L, unsigned newline)
 
         switch (type) {
             case LUA_TNUMBER:
+            num = (double) lua_tonumber(L, i);
```
Comment (agentzh) [inline src/ngx_http_lua_output.c:98]: Bad indentation?

File: src/ngx_http_lua_output.c [hunk 5]
```diff
@@ -173,6 +184,28 @@ ngx_http_lua_ngx_echo(lua_State *L, unsigned newline)
...
+                                 "%.14g", num);
+                    if (n < 0) {
+                        ngx_log_error(NGX_LOG_ERR, ngx_cycle->log, ngx_errno,
+                                      "snprintf(\"%f\") failed");
+
+                    } else {
+                        b->last += n;
+                    }
+                }
+                break;
```
Comment (agentzh) [inline src/ngx_http_lua_output.c:207]: Always add a blank line after code blocks.

File: src/ngx_http_lua_socket_udp.c [hunk 6]
```diff
@@ -865,6 +895,7 @@ ngx_http_lua_socket_udp_send(lua_State *L)
             return luaL_error(L, "impossible to reach here");
     }
 
+    query.len = p - query.data;
```
Comment (agentzh) [inline src/ngx_http_lua_socket_udp.c:898]: We should use `lua_assert()` to make sure the length is not exceeding the original upper limit `len`, just like the cosocket send code.

File: src/ngx_http_lua_output.c (hunk 5 above)
Comment (agentzh) [inline src/ngx_http_lua_output.c:207]: Still not resolved?

File: src/ngx_http_lua_output.c (hunk 4 above)
Comment (agentzh) [inline src/ngx_http_lua_output.c:98]: Still not resolved?

File: t/009-log.t [hunk 7]
```diff
@@ -548,3 +548,23 @@ ok
...
 --- error_log eval
 "2: hello\0world, client: "
+
+
+
+=== TEST 27: test log-level STDERR
+Note: maximum number of digits after the decimal-point character is 13
+--- config
+    location /log {
+        content_by_lua '
```
Comment (thibaultcha) [inline t/009-log.t:558]: We should use the new `content_by_lua_block` directive in new test cases.

File: src/ngx_http_lua_socket_tcp.c [hunk 8]
```diff
@@ -2793,13 +2806,35 @@ ngx_http_lua_socket_tcp_send(lua_State *L)
...
+                 */
+                n = snprintf((char *) b->last, NGX_DOUBLE_LEN, "%.14g", num);
+                if (n < 0) {
+                    ngx_log_error(NGX_LOG_ERR, ngx_cycle->log, ngx_errno,
+                                  "snprintf(\"%f\") failed");
+
+                } else {
+                    b->last += n;
+                }
+            }
```
Comment (thibaultcha) [inline src/ngx_http_lua_socket_tcp.c:2827]: Couldn't we avoid so much code duplication by making the len calculation and the buffer write into two inlined functions instead?

File: src/ngx_http_lua_output.h [hunk 9]
```diff
@@ -22,6 +22,47 @@ u_char *ngx_http_lua_copy_str_in_table(lua_State *L, int index, u_char *dst);
...
+ngx_http_lua_write_num(lua_State *L, int idx, u_char *dst)
+{
+    double     num;
+    int        n;
+
+    num = (double) lua_tonumber(L, idx);
+    if (num == (double) (long) num) {
+        dst = ngx_snprintf(dst, NGX_INT64_LEN, "%l", (long) num);
+
+        } else {
```
Comment (agentzh) [inline src/ngx_http_lua_output.h:49]: Bad indentation. Be careful.

File: src/ngx_http_lua_output.h [hunk 10]
```diff
@@ -22,6 +22,47 @@ u_char *ngx_http_lua_copy_str_in_table(lua_State *L, int index, u_char *dst);
...
+            /**
+             * The maximum number of significant digits is 14 in lua.
+             * Please refer to lj_strfmt.c:411 for more details.
+             */
+            n = snprintf((char *) dst, NGX_DOUBLE_LEN, "%.14g", num);
+            if (n < 0) {
+                ngx_log_error(NGX_LOG_ERR, ngx_cycle->log, ngx_errno,
+                              "snprintf(\"%f\") failed");
+
+        } else {
```
Comment (agentzh) [inline src/ngx_http_lua_output.h:59]: Ditto.

File: src/ngx_http_lua_output.c [hunk 11]
```diff
@@ -301,6 +307,10 @@ ngx_http_lua_calc_strlen_in_table(lua_State *L, int index, int arg_i,
 
         switch (type) {
             case LUA_TNUMBER:
+
```
Comment (agentzh) [inline src/ngx_http_lua_output.c:310]: Let's remove this extra blank line.

File: src/ngx_http_lua_output.h [hunk 12]
```diff
@@ -22,6 +22,48 @@ u_char *ngx_http_lua_copy_str_in_table(lua_State *L, int index, u_char *dst);
 ngx_int_t ngx_http_lua_flush_resume_helper(ngx_http_request_t *r,
     ngx_http_lua_ctx_t *ctx);
```
Comment (agentzh) [inline src/ngx_http_lua_output.h:24]: Always use 2 blank lines to separate.

File: src/ngx_http_lua_output.h [hunk 13]
```diff
@@ -22,6 +22,48 @@ u_char *ngx_http_lua_copy_str_in_table(lua_State *L, int index, u_char *dst);
...
+
+    num = (double) lua_tonumber(L, idx);
+    if (num == (double) (long) num) {
+        return NGX_INT64_LEN;
+
+    } else {
+        return NGX_DOUBLE_LEN;
+    }
+}
+
```
Comment (agentzh) [inline src/ngx_http_lua_output.h:39]: Ditto and other similar places.

File: src/ngx_http_lua_output.h [hunk 14]
```diff
@@ -22,6 +22,48 @@ u_char *ngx_http_lua_copy_str_in_table(lua_State *L, int index, u_char *dst);
...
+static ngx_inline size_t
+ngx_http_lua_get_num_len(lua_State *L, int idx)
+{
+    double     num;
+
+    num = (double) lua_tonumber(L, idx);
+    if (num == (double) (long) num) {
+        return NGX_INT64_LEN;
+
+    } else {
```
Comment (agentzh) [inline src/ngx_http_lua_output.h:35]: The previous branch never falls through and we should save this `else`.

File: src/ngx_http_lua_output.h [hunk 15]
```diff
@@ -23,6 +23,50 @@ ngx_int_t ngx_http_lua_flush_resume_helper(ngx_http_request_t *r,
...
+    num = (double) lua_tonumber(L, idx);
+    if (num == (double) (long) num) {
+        dst = ngx_snprintf(dst, NGX_INT64_LEN, "%l", (long) num);
+
+    } else {
+        /**
+         * The maximum number of significant digits is 14 in lua.
+         * Please refer to lj_strfmt.c:411 for more details.
+         */
+        n = snprintf((char *) dst, NGX_DOUBLE_LEN, "%.14g", num);
```
Comment (spacewander) [inline src/ngx_http_lua_output.h:56]: Since this PR is for performance, would it be better to copy and use LuaJIT's `lj_strfmt_num` implementation?

Comment (zhuizhuhaomeng) [reply src/ngx_http_lua_output.h:56]: if wo would like to copy LuaJIT's lj_strfmt_num implementation, we should do a performance test first.
So don't change just now.

**Pull request merged**